cpp/src/copying/gather.cuh

```diff
@@ -0,0 +1,352 @@
+    gdf_size_type const num_destination_rows{destination_column->size};
+
+    // If gathering in-place or scattering nvstring
+    // (in which case the sync_nvstring_category should be set to true)
+    // allocate temporary buffers to hold intermediate results
+    bool const sync_category =
+      std::is_same<column_type, cudf::nvstring_category>::value &&
+      sync_nvstring_category;
+    bool const in_place = !sync_category && (source_data == destination_data);
+
+    gdf_column temp_src{};
+    gdf_column temp_dest{};
+
+    if (sync_category) {
+      // sync the categories.
+      temp_src = cudf::copy(*source_column);
+      temp_dest = cudf::copy(*destination_column);
+      const gdf_column *const input_columns[2] = {source_column, &temp_dest};
+      gdf_column *output_columns[2] = {&temp_src, destination_column};
+
+      CUDF_EXPECTS(GDF_SUCCESS ==
+		   sync_column_categories(input_columns, output_columns, 2),
+		   "Failed to synchronize NVCategory");
+
+      source_data = static_cast<column_type *>(temp_src.data);
+    }
+
+    rmm::device_vector<column_type> in_place_buffer;
+    if (in_place) {
+      in_place_buffer.resize(num_destination_rows);
+      destination_data = in_place_buffer.data().get();
+    }
+
+    if (ignore_out_of_bounds) {
+      thrust::gather_if(rmm::exec_policy(stream)->on(stream), gather_map,
+			gather_map + num_destination_rows, gather_map,
+			source_data, destination_data,
+			bounds_checker<decltype(*gather_map)>{0, source_column->size});
+    } else {
+      thrust::gather(rmm::exec_policy(stream)->on(stream), gather_map,
+		     gather_map+num_destination_rows, source_data,
+		     destination_data);
+    }
+
+    if (sync_category) {
+      gdf_column_free(&temp_src);
+      gdf_column_free(&temp_dest);
+    }
+
+    // Copy temporary buffers used for in-place gather to destination column
+    if (in_place) {
+      thrust::copy(rmm::exec_policy(stream)->on(stream), destination_data,
+		   destination_data + num_destination_rows,
+		   static_cast<column_type *>(destination_column->data));
+    }
+
+    CHECK_STREAM(stream);
+  }
+};
+
+/**---------------------------------------------------------------------------*
+* @brief Function object for applying a transformation on the gathermap
+* that converts negative indices to positive indices
+*
+* A negative index `i` is transformed to `i + size`, where `size` is
+* the number of elements in the column being gathered from.
+* Allowable values for the index `i` are in the range `[-size, size)`.
+* Thus, when gathering from a column of size `10`, the index `-1`
+* is transformed to `9` (i.e., the last element), `-2` is transformed
+* to `8` (the second-to-last element) and so on.
+* Positive indices are unchanged by this transformation.
+*---------------------------------------------------------------------------**/
+template <bool enable, typename map_type>
+struct negative_index_converter : public thrust::unary_function<map_type,map_type>{};
```

Comment:
Double negatives can be confusing. 
How about making this an `index_converter` and changing `enable` to be `negative`?
Seems it would be clearer to enable negative on something positive than disabling negative to make something positive.

Reply:
Agree that it can be confusing. How about an `enum` template parameter that makes it more explicit what the converter does?

```
enum index_conversion {
    NEGATIVE_TO_POSITIVE = 0,
    SOMETHING_ELSE =1 ,
    NONE = 2,
}
```

Reply:
Yes. I saw this line and had to look up what it was doing.
```
negative_index_converter<false,map_type>{...}
```
Something like this perhaps:
```
template <typename map_type, bool enable = NOTHING>
struct index_converter ...
```
and then maybe
```
index_converter<map_type, NEGATIVE>{ ... }  
```
and normal pass through would be just
```
index_converter<map_type>{ ... }  
```